### ISSUE TITLE
training speed of BN-Inception

### ISSUE THREAD

**Issue by abangdd** [apache/incubator-mxnet#1424]
the training speed of BN-Inception is 10,666 seconds per epoch for one TitanX
see https://github.com/dmlc/mxnet/tree/master/example/image-classification
while in Torch, the training speed is 7686 seconds per epoch (192 ms per 32 images)
see http://torch.ch/blog/2016/02/04/resnets.html
the difference is not neglectable, do you know the reasons?

**Comment by piiswrong** [comment]
Did you run the tests to see the current number? The doc is probably out dated since it was likely produced with cudnn v3.
@tqchen @jermainewang

**Comment by mli** [comment]
that's interesting, probably due to cudnn v3/v4 or the fft-based convolution. we did measurement a month ago, these platforms, torch/caffe/mxnet, are comparable on cnns. we need to find out what happens there.

**Comment by abangdd** [comment]
one reason I guess is the usage of cudnn batchnormalization routine in torch
as we know, the cuda libraries are substantially faster than nvcc compiled cuda codes

**Comment by mavenlin** [comment]
There is one commit in between Jan 22 and Feb 7 that slows down the speed... Trying to find that.

**Comment by ajtulloch** [comment]
FWIW, cudnn's algorithm selection as of v3/v4 can be suboptimal (even for common convnets like VGG-16), so using the 'Find' variant (which uses an autotuner to select the algorithm) is generally recommended for maximum performance vs the 'Get' variant (which MXNet uses, and which can result in suboptimal algorithm choices). See e.g. https://github.com/soumith/cudnn.torch/blob/28c2f6e76a0d3671ce127197c25e39c5ee4be627/SpatialConvolution.lua#L186-L190

I'm not sure on which BN implementation was being reported there, but IIRC the FB SpatialBatchNormalization GPU implementation was faster than (at least the initial) cudnn BatchNormalization implementation.

**Comment by piiswrong** [comment]
@ajtulloch Thanks for pointing this out. One problem is MXNet pre-allocates all memory to do memory optimization. This function seems to allocate memory internally and need to be called before user memory allocation. @tqchen @antinucleon Any ideas?

**Comment by ajtulloch** [comment]
I'm not sure if it's the case here FWIW, but it might be worth just logging which algorithms are chosen by `Get` vs `Find` (even just running the Torch version with cudnn.benchmark = true/false) and making sure MXNet is choosing the right kernels. IIRC MXNet uses multiple host threads to launch kernels on multiple devices (right?), so it doesn't have to add the Torch nn.DataParallelTable multi-threading hack needed to overcome the launch overhead.

**Comment by tqchen** [comment]
@piiswrong  The temp space in mxnet can be dynamically resized

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
